src/diffpy/srmise/basefunction.py

```diff
@@ -103,9 +103,9 @@ def __init__(
             if type(p) not in (str, unicode):
                 emsg = "Argument parameterdict's keys must be strings."
                 raise ValueError(emsg)
-        vals = self.parameterdict.values()
-        vals.sort()
-        if vals != range(self.npars):
+        expected_values = set(range(self.npars))
```

Comment:
This is a slightly different test from the original. That may be ok, but let's make sure.  Also the error message is pretty incomprehensible so I don't understand what the code is actually doing. Could you try rewriting that emsg based on your understanding?

Reply:
This change looks good to me.  Using `set()` fulfills the same purpose as the `sort()` in the original code, and it is good to make what is expected explicit.

As for what this code does, it checks that the indices specified by a particular parameterization of a mathematical function are what is expected for a tuple of the same length.  This is intended as an aid to someone implementing a new peak or baseline -- once that class has been written correctly it should always pass.  For example, gaussian.py defines
```
        parameterdict={'position':0,'width':1,'area':2}
```
A Gaussian has three three parameters, so the expected indices are 0, 1, and 2.  The actual set of values in the code above are also 0, 1, and 2.  Since these are the same the test passes.  If the code had been `{'area':2,'width':1,'position':0}` that would be OK, it is equivalent to the original dictionary.  However, `{'position':0,'width':1,'area':20}` would produce an error, because a sequence of length three does not have a 20th element.  And `{'position':0, 'width':0,'area':2}` would produce an error because the index 1 is missing and because any sequence has exactly one element with a given index.

This check doesn't detect all possible errors.  For example,  `{'position':2,'width':1,'area':0}` would pass the test, but would contain a semantic error because the 0th element in my implementation of the Gaussian class is actually position, not area.

I'll meditate on a more useful error message, I agree it is not good.

Reply:
> This change looks good to me. Using `set()` fulfills the same purpose as the `sort()` in the original code, and it is good to make what is expected explicit.
> 
> As for what this code does, it checks that the indices specified by a particular parameterization of a mathematical function are what is expected for a tuple of the same length. This is intended as an aid to someone implementing a new peak or baseline -- once that class has been written correctly it should always pass. For example, gaussian.py defines
> 
> ```
>         parameterdict={'position':0,'width':1,'area':2}
> ```
> 
> A Gaussian has three three parameters, so the expected indices are 0, 1, and 2. The actual set of values in the code above are also 0, 1, and 2. Since these are the same the test passes. If the code had been `{'area':2,'width':1,'position':0}` that would be OK, it is equivalent to the original dictionary. However, `{'position':0,'width':1,'area':20}` would produce an error, because a sequence of length three does not have a 20th element. And `{'position':0, 'width':0,'area':2}` would produce an error because the index 1 is missing and because any sequence has exactly one element with a given index.
> 
> This check doesn't detect all possible errors. For example, `{'position':2,'width':1,'area':0}` would pass the test, but would contain a semantic error because the 0th element in my implementation of the Gaussian class is actually position, not area.
> 
> I'll meditate on a more useful error message, I agree it is not good.

Set is less strict than sort.  That was my point.

Reply:
I think normally these days we would handle these things with tests rather than dynamic code checks.  @stevenhua0320 would you like to design a test and we can get rid of this in the code?

Reply:
Ah, that's a good point.  I don't think it matters in this case because the dictionary is already unordered.  And the case where set drops a duplicate index will still cause an error because the number of parameters is wrong.  For example,
```
parameterdict={'position':0,'width':1,'area':2,`meaningless`:1}
```
would produce the actual set {0,1,2}, but the expected set would be {0,1,2,3} and so it would appropriately fail.